src/control/src/point_solver/psolver.cpp

```diff
@@ -57,56 +56,89 @@ std::pair<Position, int> PointSolver::update_closest_point(
-    // ld if it's not withing ld margin), check if it is the closest point found yet
-    if (distance > 1 && std::abs(ld - distance) < closest_distance) {
-      RCLCPP_INFO(rclcpp::get_logger("control"), "New closest point found that doesnt fint margin: %f", distance);
-      closest_distance = std::abs(ld - distance);
-      closest_yet = pathpoint_array[index];
+    double result_x;
+    double result_y;
+
+    // Slope of the line is infinite, don't need line equation
+    if (point_a.x == point_b.x) {
+      RCLCPP_DEBUG(rclcpp::get_logger("control"), "Vertical line!!");
+      result_x = point_a.x;
+      double delta = ld * ld - std::pow(result_x - rear_axis_point.x, 2);
+      if (delta < 0) {
+        continue;
+      }
+      double y1 = rear_axis_point.y + std::sqrt(delta);
+      double y2 = rear_axis_point.y - std::sqrt(delta);
+
+      if (y1 >= std::min(point_a.y, point_b.y) && y1 <= std::max(point_a.y, point_b.y)) {
+        result_y = y1;
+      } else {
+        result_y = y2;
+      }
+
+    } else {
+      // y = mx + c
+      double m = (point_b.y - point_a.y) / (point_b.x - point_a.x);
+      double c = point_a.y - m * point_a.x;
+
+      // (x - x0)^2 + (y - y0)^2 = r^2
+      // with x0 = rear_axis_point.x, y0 = rear_axis_point.y, r = ld
+
+      // with this information we can find the intersection point between the circle and the line
+      // and obtain the lookahead point.
+      // Expanded form of the circle equation with y substituted by mx + c gives a quadratic
+      // equation:
+      double A = 1 + m * m;
+      double B = 2 * (m * c - rear_axis_point.x - m * rear_axis_point.y);
+      double C = rear_axis_point.x * rear_axis_point.x + c * c +
+                 rear_axis_point.y * rear_axis_point.y - 2 * c * rear_axis_point.y - ld * ld;
+
+      double delta = B * B - 4 * A * C;
+
+      if (delta < 0) {
+        continue;
+      }
+
+      double x1 = (-B + std::sqrt(delta)) / (2 * A);
+      double x2 = (-B - std::sqrt(delta)) / (2 * A);
+
+      double y1 = m * x1 + c;
+      double y2 = m * x2 + c;
+
+      // solution (x,y) which is within bounds of point_a and point_b
+      if (x1 >= std::min(point_a.x, point_b.x) && x1 <= std::max(point_a.x, point_b.x) &&
+          y1 >= std::min(point_a.y, point_b.y) && y1 <= std::max(point_a.y, point_b.y)) {
+        result_x = x1;
+        result_y = y1;
+      } else {
+        result_x = x2;
+        result_y = y2;
+      }
     }
-  }
-  RCLCPP_INFO(rclcpp::get_logger("control"), "Rear axis point: x: %f, y: %f", rear_axis_point.x, rear_axis_point.y);
-
 
-  if (closest_distance == std::numeric_limits<double>::max()) {
-    RCLCPP_WARN(rclcpp::get_logger("control"), "No lookahead point found");
-    return std::make_tuple(Position(), 0.0, true);
+    return std::make_tuple(Position(result_x, result_y),
+                           (pathpoint_array[index_a].v + pathpoint_array[index_b].v) / 2.0, false);
```

Comment:
Why is the velocity also the arithmetic mean of both velocities instead of doing a linear interpolation?

Reply:
because the changes should be like a meter apart and already quite smooth. besides, planning atm sends constant speed at every pathpoint. this could be nice to have implemented later though, to have the loweest possible difference in speed metrics